base/locks-mt.jl

```diff
@@ -28,15 +28,38 @@ a lock is the wrong way to synchronize.
 See also [`Mutex`](@ref) for a more efficient version on one core or if the
 lock may be held for a considerable length of time.
 """
-struct SpinLock <: AbstractLock
-    handle::Atomic{Int}
-    SpinLock() = new(Atomic{Int}(0))
-end
+mutable struct SpinLock <: AbstractLock
+    handle::Int
+    SpinLock() = new(0)
+end
+
+import Base.Sys.WORD_SIZE
+
+@eval _xchg!(x::SpinLock, v::Int) =
+    llvmcall($"""
```

Comment:
Oh evil! This seems fine but, we do need a better interface for atomics (I am not that big of fan of the use of the string-based llvmcall interface).

I quite like what TIm came up with in https://github.com/JuliaGPU/CUDAnative.jl/pull/370 and I recall us discussing `RefField` to obtain a pointer to a field in a mutable struct.

Reply:
Yes, very nice. I think `@atomic` could be in Base and we could use RefArray and RefField for this.